bugsnag-android-core/src/main/java/com/bugsnag/android/InternalMetrics.kt

```diff
@@ -0,0 +1,92 @@
+package com.bugsnag.android
+
+import java.util.Collections
+
+/**
+ * Stores internal metrics for Bugsnag use.
+ */
+internal class InternalMetrics {
+    private var enabled = false
+    private val configDifferences = mutableMapOf<String, Any>()
+    private val callbackCounts = mutableMapOf<String, Int>()
+
+    fun applyTelemetryConfig(telemetry: Set<Telemetry>) {
+        enabled = telemetry.contains(Telemetry.USAGE)
+        NdkPluginCaller.setInternalMetricsEnabled(enabled)
+    }
+
+    fun toJsonableMap(): Map<String, Any> {
+        return if (enabled) {
+            mapOf(
+                "config" to configDifferences,
+                "callbacks" to allCallbacks()
+            )
+        } else {
+            mapOf()
+        }
+    }
+
+    fun setConfigDifferences(differences: Map<String, Any>) {
+        if (enabled) {
+            configDifferences.clear()
+            configDifferences.putAll(differences)
+            NdkPluginCaller.setStaticData(Collections.singletonMap("config", configDifferences))
+        }
+    }
+
+    fun setCallbackCounts(newCallbackCounts: Map<String, Int>) {
+        if (enabled) {
+            callbackCounts.clear()
+            callbackCounts.putAll(newCallbackCounts)
+            NdkPluginCaller.initCallbackCounts(newCallbackCounts)
+        }
+    }
+
+    fun notifyAddCallback(callback: String) {
+        if (enabled) {
+            modifyCallback(callback, 1)
+            NdkPluginCaller.notifyAddCallback(callback)
+        }
+    }
+
+    fun notifyRemoveCallback(callback: String) {
+        if (enabled) {
+            modifyCallback(callback, -1)
+            NdkPluginCaller.notifyRemoveCallback(callback)
+        }
+    }
+
+    private fun modifyCallback(callback: String, delta: Int) {
+        var currentValue = callbackCounts[callback]
+        if (currentValue == null) {
+            currentValue = delta
+        } else {
+            currentValue += delta
+        }
+        if (currentValue < 0) {
+            currentValue = 0
+        }
+        callbackCounts[callback] = currentValue
+    }
+
+    private fun allCallbacks(): Map<String, Any> {
+        val result = mutableMapOf<String, Any>()
+        result.putAll(callbackCounts)
+
+        val counts = NdkPluginCaller.getCurrentCallbackSetCounts()
```

Comment:
The NDK plugin doesn't call Java to update the counts on each native API call. It's polled from the Java side as-needed in order to avoid slowing things down too much.

Reply:
That could still be done if the NDK plugin module was responsible for providing those counts without reflection, possibly using a similar technique to how we allow plugins - each module that is active would add a callback to be invoked here:

```kotlin
metricProviders.forEach { provider ->
    val counts = provider.currentCallbackSetCounts()
    ...
}
```

That would also account for metrics added by ReactNative, Flutter and Unreal

Reply:
Are we anticipating needing such metrics in the near future? Otherwise I'd prefer to keep a braindead implementation until we actually need more. Changing it over later won't be much work.